### PR TITLE
Introduce Hibernate naming strategies

### PR DESCRIPTION
**Pull request by buehner**
Regarding #93 , this PR introduces a DDL naming strategy for SHOGun2:

Since Hibernate 5, the naming strategy is split up in two parts:

* *ImplicitNamingStrategy* - If an explicit naming configuration is given via `@Table` (or `@JoinTable`), this strategy will **NOT** be executed. Otherwise: Based on the class name, an implicit name for the entity will be created. In our case, this strategy will produce *plural forms* of the singular class name forms, e.g. `Application -> Applications`. 
* *PhysicalNamingStrategy* - This strategy will **always** be executed: Based on the result of the implicit naming strategy or an explicit naming configuration, this strategy will create a physical name for tables/schemas etc.  In our case, the implementation will
  1. transform all entity names to **lower case**
  2. **limit the length of an identifier** according to the given JDBC context (currently H2, Oracle and PostgreSQL are supported)

As a result of this implementation, a lot of `@Table` and `@JoinTable` annotations have been cleaned up. Even though, in case of most `@JoinTable`s the naming should still be done manually to avoid that the physical naming strategy cuts off the (implicitly) generated join table names (including `_`) in case of a limitation.

Tests are included!

### CODE REVIEW

**Comment by marcjansen** [comment]
This is looking very neat, @buehner!

How do you configure a prefix for all tables, be they explicitly named or generated from a model? This is missing, right?

I imagen that an application can create a bean / instance of `ImplicitNamingStrategyShogun2` and give it the prefix to use.

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java [hunk 1]
```diff
@@ -43,7 +43,6 @@
 
 	@ManyToMany
 	@JoinTable(
-		name = "USER_ROLE",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java]: This will become `users_roles`, right?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/map/MapControl.java [hunk 2]
```diff
@@ -50,15 +50,15 @@
 	 *
 	 */
 	@ElementCollection
-	@MapKeyColumn(name = "MAPCTRLPROPERTY")
+	@MapKeyColumn(name = "PROPERTY")
 	@Column(name = "VALUE")
 	@CollectionTable(
-			name = "MAPCONTROL_MAPCTRLPROPERTIES",
-			joinColumns = @JoinColumn(name = "MAPCONTROL_ID")
+		name = "MAPCONTROLS_PROPERTIES",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/map/MapControl.java]: Why do you explicitly give a name here, bit not everywhere else?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/XyzLayerDataSource.java [hunk 3]
```diff
@@ -43,7 +43,7 @@
 	private Extent extent;
 
 	@OneToMany
-	@JoinTable(name = "LAYERDATASOURCE_RESOLUTIONS")
+	@JoinTable(name = "XYZLAYERDATASRC_RESOLUTIONS")
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/XyzLayerDataSource.java]: Why do you explicitly give a name here, but not everywhere else?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java [hunk 4]
```diff
@@ -63,9 +63,9 @@
 
 	@ManyToMany
 	@JoinTable(
-			name = "TILEWMSLAYERDATASOURCE_STYLE",
-			joinColumns = { @JoinColumn(name = "TILEWMSLAYERDATASOURCE_ID") },
-			inverseJoinColumns = { @JoinColumn(name = "STYLE_ID") }
+		name = "TILEWMSLAYERDATASOURCE_STYLE",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java]: Why do you explicitly give a name here, but not everywhere else?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java [hunk 5]
```diff
@@ -46,9 +46,9 @@
 
 	@ManyToMany
 	@JoinTable(
-			name = "TILEWMSLAYERDATASRC_LAYERNAME",
-			joinColumns = { @JoinColumn(name = "TILEWMSLAYERDATASOURCE_ID") },
-			inverseJoinColumns = { @JoinColumn(name = "LAYERNAME_ID") }
+		name = "TILEWMSLAYERDATASRC_LAYERNAME",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java]: Why do you explicitly give a name here, but not everywhere else?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/ImageWmsLayerDataSource.java [hunk 6]
```diff
@@ -44,34 +44,34 @@
...
 	)
 	@JsonIdentityReference(alwaysAsId = true)
 	private List<GeoWebServiceLayerName> layerNames;
 
 	@ManyToMany
 	@JoinTable(
-			name = "IMAGEWMSLAYERDATASOURCE_STYLE",
-			joinColumns = { @JoinColumn(name = "IMAGEWMSLAYERDATASOURCE_ID") },
-			inverseJoinColumns = { @JoinColumn(name = "STYLE_ID") }
+		name = "IMAGEWMSLAYERDATASOURCE_STYLE",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/ImageWmsLayerDataSource.java]: Why do you explicitly give a name here, but not everywhere else?

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/CompositeModule.java [hunk 7]
```diff
@@ -54,9 +54,9 @@
 	@ManyToMany
 	@Cascade(CascadeType.SAVE_UPDATE)
 	@JoinTable(
-			name = "MODULE_SUBMODULE",
-			joinColumns = { @JoinColumn(name = "MODULE_ID") },
-			inverseJoinColumns = { @JoinColumn(name = "SUBMODULE_ID") }
+		name = "MODULE_SUBMODULE",
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/CompositeModule.java]: Why do you explicitly give a name here, but not everywhere else? Also shouldn't this be `MODULES_SUBMODULES``

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java [hunk 8]
```diff
@@ -54,11 +54,10 @@
 	/**
 	 * The layers used within this OverviewMap.
 	 */
-	@ElementCollection
+	@ManyToMany
```
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java]: Why is this change needed?

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+ */
+public class ImplicitNamingStrategyShogun2 extends ImplicitNamingStrategyJpaCompliantImpl {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final String PLURAL_SUFFIX_S = "s";
+
+	private static final String LAST_CHAR_S = "s";
+	private static final String LAST_CHAR_X = "x";
+	private static final String LAST_CHAR_Z = "z";
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: Shoukld the above four constants probably be of type `char`?

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	private static final String LAST_CHARS_CH = "ch";
+	private static final String LAST_CHARS_SH = "sh";
+	private static final String PLURAL_SUFFIX_ES = "es";
+
+	private static final String LAST_CHAR_Y = "y";
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: Should this be type `char`?

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+	 * @return
+	 */
+	private String transformToPluralForm(String singular) {
+		StringBuilder plural = new StringBuilder();
+
+		// start with singular form
+		plural.append(singular);
+
+		if (singular.endsWith(LAST_CHAR_Y)) {
+			// replace last char with suffix form
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: Perhaps an inline example helps future readers:

```java
// e.g. "city" => "cities"
```

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+			int length = plural.length();
+			plural.replace(length - 1, length, PLURAL_SUFFIX_IES);
+
+		} else if (singular.endsWith(LAST_CHAR_S)
+				|| singular.endsWith(LAST_CHAR_X)
+				|| singular.endsWith(LAST_CHAR_Z)
+				|| singular.endsWith(LAST_CHARS_CH)
+				|| singular.endsWith(LAST_CHARS_SH)) {
+
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: This could also benefit from inilne examples (takle the values from the tests)

**Comment by marcjansen** [comment]
I have added some minor inline comments and on top of the inline diffs some other qquestions. Thanks for the work on this.

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java (hunk 1 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java]: Correct!

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/map/MapControl.java (hunk 2 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/map/MapControl.java]: Because the implicitly generated value in this case would exceed 30 chars (but the PhysicalNamingStrategy cuts off after 30 chars in case of oracle)

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/XyzLayerDataSource.java (hunk 3 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/XyzLayerDataSource.java]: see above...

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java (hunk 4 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java]: see above...

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java (hunk 5 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/TileWmsLayerDataSource.java]: see above...

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/ImageWmsLayerDataSource.java (hunk 6 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/layer/source/ImageWmsLayerDataSource.java]: see above...

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/CompositeModule.java (hunk 7 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/CompositeModule.java]: Because `MODULES_SUBMODULES` is nicer than `COMPOSITEMODULES_MODULES` ;-) I will correct this to the plural form in a follow up commit

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java (hunk 8 above)
Comment (buehner) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java]: `@ElementCollection` should only be used for basic or embedabble types, which was not correct here. It seems that not all of the config was respected in case of `@ElementCollection`

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java (hunk 9 above)
Comment (buehner) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: Why not!? I will change this in a follow up commit

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java (hunk 10 above)
Comment (buehner) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: see above

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java (hunk 11 above)
Comment (buehner) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: Good point! I will add some inline comments

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java (hunk 12 above)
Comment (buehner) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/ImplicitNamingStrategyShogun2.java]: see above

**Comment by buehner** [comment]
To answer your questions, @marcjansen :

* You're right. I simply forgot the prefix. I will add this in a follow up commit
* Currently, our strategy only applies to names of tables/join tables.
* Here you can see the possibilities we have regarding the implicit strategy: https://docs.jboss.org/hibernate/orm/5.0/javadocs/org/hibernate/boot/model/naming/ImplicitNamingStrategy.html
* Here the doc for the physical naming strategy: https://docs.jboss.org/hibernate/orm/5.0/javadocs/org/hibernate/boot/model/naming/PhysicalNamingStrategy.html

Regarding your comments, i will add some commits soon. Thx!

**Comment by buehner** [comment]
I pushed some new commits that address your notes. Most important change: columns will now also be limited and lower case!

Feel free to merge, if you are happy. Thx

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java (hunk 1 above)
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/User.java]: OK. then I understood the concept :-)

File: src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java (hunk 8 above)
Comment (marcjansen) [inline src/shogun2-core/shogun2-model/src/main/java/de/terrestris/shogun2/model/module/OverviewMap.java]: OK

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/PhysicalNamingStrategyShogun2.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+	private static final long serialVersionUID = 1L;
+
+	protected static final int LENGTH_LIMIT_ORACLE = 30;
+
+	protected static final int LENGTH_LIMIT_POSTGRESQL = 63;
+
+	@Autowired(required = false)
+	@Qualifier("tablePrefix")
+	private String tablePrefix;
```
Comment (marcjansen) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/PhysicalNamingStrategyShogun2.java]: Why did you chose

```java
private String tablePrefix;
```

over

```java
private String tablePrefix = null;
```

or even

```java
private String tablePrefix = "";
```

are these equivalent?

**Comment by marcjansen** [comment]
I have added some more comments. Can you pointme to a test where the prefix behaviour is tested?

Also (but probably better done in a follow-up PR) I would like to limit the prefix to max 4 characters, with only a subset of allowe chars: `/^[a-z_]{0,4}/$`. What do you think?

File: src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/PhysicalNamingStrategyShogun2.java (hunk 13 above)
Comment (buehner) [inline src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/PhysicalNamingStrategyShogun2.java]: At first, the `@Autowired` annotation is configured with `required=false`, i.e. the prefix value is not mandatory.

`private String tablePrefix;` is the equal to `private String tablePrefix = null;` as `null` is the default value.

On top of that, the prefix value will only be used if it is not null:

https://github.com/buehner/shogun2/blob/naming-strategy/src/shogun2-util/src/main/java/de/terrestris/shogun2/util/naming/PhysicalNamingStrategyShogun2.java#L69-L71

**Comment by buehner** [comment]
@marcjansen I added three tests for the table prefix in 82a1b42

Please review and merge. Thx

**Comment by marcjansen** [comment]
Thanks. Good to go.
